chore(sql): off-heap and flyweight histogram

mcadariu opened the pull request:
Related to https://github.com/questdb/questdb/issues/4120. 

The implementation is inspired by the existing on-heap `Histogram`, and the existing `GroupByArraySink`.

The tests compare outputs to the on-heap `Histogram` which is the reference implementation / ground truth.

mcadariu committed:
Add GroupByHistogramSink

mcadariu committed:
Method renaming

mcadariu committed:
Make histogram flyweight

mcadariu committed:
cleanup comments

mcadariu committed:
split up the resize method into some helpers

mcadariu committed:
Remove cached fields

mcadariu committed:
Rearrange private static final fields

mcadariu committed:
Method renaming

mcadariu committed:
implement merge method and tests

mcadariu commented:
The `Rust Test and Lint` job seems to be failing, should be unrelated

puzpuzpuz commented on core/src/main/java/io/questdb/std/histogram/org/HdrHistogram/GroupByHistogramSink.java:
```diff
@@ -0,0 +1,381 @@
+    }
+
+    @Override
+    public GroupByHistogramSink copy() {
```
I'd rather not bother with the methods we don't need and simply throw `UnsupportedOperationException`. Same applies to `copyCorrectedForCoordinatedOmission` and maybe some other methods here. We only need the minimal required methods.

mcadariu replied:
Ah, righjt! A bit of a by-product of me working in a kind of bottom-up fashion, indeed we should have nothing but what is strictly necessary.

mcadariu committed:
apply review comments

mcadariu committed:
add timestamps to the header (used in merge) and a test for repointing histograms

mcadariu committed:
update test with timestamp handling checks

puzpuzpuz commented on core/src/main/java/io/questdb/std/histogram/org/HdrHistogram/GroupByHistogramSink.java:
```diff
@@ -0,0 +1,381 @@
+ * <pre>
+ * | totalCount | normalizingIndexOffset | maxValue | minNonZeroValue | histogram counts array |
+ * +------------+------------------------+----------+-----------------+------------------------+
+ * |  8 bytes   |        4 bytes         | 8 bytes  |    8 bytes      | countsArrayLength * 8  |
```
> Am _I_ missing something? 🤔 If you think it's best I can definitely remove the `extends AbstractHistogram` and see what happens.

I'm just seeing a lot of fields (think, state) in that hierarchy of classes while we need this object to be an off-heap flyweight, i.e. its state is fully loaded from off-heap memory when you make an `of()` call.

mcadariu replied:
Ah pff, there's fields in `AbstractHistogramBase` too 🤦
I am changing the approach by indeed not extending `AbstractHistogram` and trying to get the tests to pass that way. Potentially the duplication can be solved by composition instead of inheritance, but I'll cross that bridge when I get there.

mcadariu committed:
GroupByHistogram without extending the AbstractHistogram

mcadariu committed:
fix field and method visibility

mcadariu committed:
remove unused fields

mcadariu committed:
update copyright year

mcadariu committed:
remove redundant timestamp variables and setter calls have been removed since add() already handles the timestamp merging

puzpuzpuz commented:
@mcadariu thanks for updating the PR! Porting `HdrHistogram` to off-heap memory is certainly tough. I should be able to do another review round later this week.

mcadariu committed:
Use Vect.memmove() to handle overlapping regions correctly.

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

mcadariu committed:
fix test method name to better match the behaviour

mcadariu commented:
@coderabbitai review

puzpuzpuz commented:
> @coderabbitai review

Not code rabbit, but still. Sorry for the delays with the review. I'll do my best to check the updates tomorrow. In the meantime, have you tried porting the `approx_percentile()` SQL function to this histogram implementation and measuring the performance before/after?

puzpuzpuz commented on core/src/main/java/io/questdb/std/histogram/org/HdrHistogram/GroupByHistogram.java:
```diff
@@ -0,0 +1,748 @@
+            this.highestTrackableValue = Unsafe.getUnsafe().getLong(ptr + highestTrackableValuePosition);
+            this.allocatedSize = headerSize + (countsArrayLength * 8L);
+
+            recalculateDerivedFields();
```
I'm afraid we need to measure if parallel `approx_percentile()` based on this class is any faster than the `HdrHistogram` one. So far, this is the most complex `GroupBy*` flyweight we have and I wonder how much overhead state loading/storing adds.

mcadariu committed:
Add fuzz test for record and merge functions

mcadariu committed:
Remove redundant methods

mcadariu committed:
Address review comments

mcadariu committed:
Add repoint-and-record fuzz test for GroupByHistogram

mcadariu committed:
Remove duplicate CairoException import

mcadariu commented on core/src/test/java/io/questdb/test/std/histogram/org/HdrHistogram/GroupByHistogramTest.java:
```diff
@@ -0,0 +1,712 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class GroupByHistogramTest extends AbstractCairoTest {
```
Good one, done

puzpuzpuz commented on core/src/main/java/io/questdb/std/histogram/org/HdrHistogram/GroupByHistogram.java:
nit: `@Override` won't hurt here.

mcadariu replied:
Done 👍

puzpuzpuz commented on core/src/main/java/io/questdb/std/histogram/org/HdrHistogram/GroupByHistogram.java:
nit: as this sink is no longer coupled with HdrHistogram library, we could move it to the `io.questdb.griffin.engine.groupby` package.

mcadariu replied:
Done 👍

puzpuzpuz reviewed (commented):
Long story short, the PR looks great. A few tests could be improved, but I'm happy to do that in a separate PR. So, let me know if you won't be able to improve the tests and I'll approve the PR and handle the tests. The description of the gaps is below.

The fuzz tests (`testRecordValueFuzz`, `testMergeFuzz`, `testRepointAndRecordFuzz`) compare against HdrHistogram and cover the core paths well. The gap is in the deterministic merge edge-case tests:

- **`testMergePreservesPercentiles`** compares a merged `GroupByHistogram` against   another `GroupByHistogram` that received all values directly. This is self-referential — if recording and merging share a systematic bug, the test still passes. The `combined` histogram should be an HdrHistogram `Histogram`.

- **`testMergeEmptyDestinationWithLargeValues`** exercises the optimized `merge()` path where `dest.ptr == 0` and the source has large values that trigger
  `resize()` instead of `ensureCapacity()`. This is the exact path changed by review comment #11. No oracle verifies the output.

- **`testMergeExtremeValueRange`** merges values spanning 1 to 9B (causing major internal resizing) but only checks `totalCount` and a rough `maxValue` bound. Percentile accuracy after such a large structural change is not verified
  against an oracle.

The deterministic tests exist and hit the right edge cases — they just assert hardcoded values rather than comparing against HdrHistogram. Adding oracle
comparisons to these three would close the gap.

mcadariu commented:
Thanks! I can address the remaining points a bit later 👍

mcadariu committed:
Move GroupByHistogram to groupby package and strengthen merge tests

Relocate GroupByHistogram and its test from std.histogram.org.HdrHistogram
to griffin.engine.groupby, and compare merge results against a Histogram
oracle for stronger assertions.

mcadariu committed:
Merge branch 'master' into off-heap-histogram-v2

puzpuzpuz commented:
@mcadariu thanks for adding the tests! The PR looks good. Linter does like a couple things, could you check? Once the linter is happy, I'll approve the PR.
```
diff --git a/core/src/main/java/io/questdb/griffin/engine/groupby/GroupByHistogram.java b/core/src/main/java/io/questdb/griffin/engine/groupby/GroupByHistogram.java
index e7be2c28e..9714d425c 100644
--- a/core/src/main/java/io/questdb/griffin/engine/groupby/GroupByHistogram.java
+++ b/core/src/main/java/io/questdb/griffin/engine/groupby/GroupByHistogram.java
@@ -352,8 +352,8 @@ public class GroupByHistogram implements Mutable {
     private void checkBounds(int index) {
         if (index < 0 || index >= countsArrayLength) {
             throw CairoException.nonCritical()
-                .put("index ").put(index)
-                .put(" out of bounds [0, ").put(countsArrayLength).put(')');
+                    .put("index ").put(index)
+                    .put(" out of bounds [0, ").put(countsArrayLength).put(')');
         }
     }
 
@@ -485,8 +485,8 @@ public class GroupByHistogram implements Mutable {
     }
 
     private int countsArrayIndex(final int bucketIndex, final int subBucketIndex) {
-        assert(subBucketIndex < subBucketCount);
-        assert(bucketIndex == 0 || (subBucketIndex >= subBucketHalfCount));
+        assert (subBucketIndex < subBucketCount);
+        assert (bucketIndex == 0 || (subBucketIndex >= subBucketHalfCount));
         int bucketBaseIndex = (bucketIndex + 1) << subBucketHalfCountMagnitude;
         int offsetInBucket = subBucketIndex - subBucketHalfCount;
         return bucketBaseIndex + offsetInBucket;
```

mcadariu committed:
Fix formatting in GroupByHistogram

mcadariu committed:
Merge branch 'master' into off-heap-histogram-v2

puzpuzpuz committed:
Merge branch 'master' into off-heap-histogram-v2

puzpuzpuz reviewed (approved):
Thanks for this epic patch!

javier commented:
I made a blog post out of this exchange. Very cool :)

https://questdb.com/blog/code-review-tripled-histogram-speedup/

mcadariu commented:
Thanks @javier, awesome write-up! I felt too that there's a story to tell here and you have nailed it.

puzpuzpuz committed:
Merge branch 'master' into off-heap-histogram-v2

bluestreak01 committed:
Merge branch 'master' into off-heap-histogram-v2
